### PR TITLE
add app version to apns_devices

### LINKED ISSUE #383: store app version to db

### PR DESCRIPTION
**Pull request by akhrail1996**
closes #383 
enabled by https://github.com/getsince/test5/commit/6565544afd245a3629e8fe7140739ff81a4916a3

### CODE REVIEW

File: lib/t_web/controllers/device_controller.ex [hunk 1]
```diff
@@ -11,6 +11,7 @@ defmodule TWeb.DeviceController do
     :ok =
       Accounts.save_apns_device_id(user_id, user_token, device_token,
         locale: params["locale"],
+        version: params["version"] || "<2.3.0",
```
Comment (akhrail1996) [inline lib/t_web/controllers/device_controller.ex:14]: 2.3.0 is the version where https://github.com/getsince/test5/commit/6565544afd245a3629e8fe7140739ff81a4916a3 will be released

**Comment by ruslandoga** [comment]
@akhrail1996 please add it to session tokens instead

**Comment by akhrail1996** [comment]
@ruslandoga ready for review

File: lib/t/accounts.ex [hunk 2]
```diff
@@ -363,13 +363,25 @@ defmodule T.Accounts do
   Gets the user with the given signed token.
   """
   def get_user_by_session_token(token, context) do
```
Comment (ruslandoga) [inline lib/t/accounts.ex:365]: please remove this function if it's not used anywhere

File: lib/t/accounts.ex [hunk 3]
```diff
@@ -363,13 +363,25 @@ defmodule T.Accounts do
...
+    get_user_by_session_token(token, nil, context)
+  end
+
+  def get_user_by_session_token(token, version, context) do
```
Comment (ruslandoga) [inline lib/t/accounts.ex:369]: please rename the function to better express it's new role

File: lib/t/accounts.ex [hunk 4]
```diff
@@ -363,13 +363,25 @@ defmodule T.Accounts do
...
       end
 
-    Repo.one(query)
+    case version do
+      nil ->
+        Repo.one(query)
+
+      _ ->
+        update = [set: [version: version]]
+        {1, users} = Repo.update_all(query, update)
```
Comment (ruslandoga) [inline lib/t/accounts.ex:382]: this would crash on tokens that don't exist in db

File: lib/t/accounts.ex [hunk 5]
```diff
@@ -363,13 +363,25 @@ defmodule T.Accounts do
...
+
+  def get_user_by_session_token(token, version, context) do
     {:ok, query} =
       case context do
         "session" -> UserToken.verify_session_token_query(token)
         "mobile" -> UserToken.verify_mobile_token_query(token)
       end
 
-    Repo.one(query)
+    case version do
```
Comment (ruslandoga) [inline lib/t/accounts.ex:376]:
```elixir
if version do
  case Repo.update_all(query, set: [version: version]) do
    {1, [user]} -> user
    {0, _} -> nil
  end
else
  Repo.one(query)
end
```

File: lib/t_web/channels/user_socket.ex [hunk 6]
```diff
@@ -19,7 +19,16 @@ defmodule TWeb.UserSocket do
       Logger.metadata(remote_ip: remote_ip)
     end
 
-    if user = Accounts.get_user_by_session_token(token, "mobile") do
+    version =
```
Comment (ruslandoga) [inline lib/t_web/channels/user_socket.ex:22]: version = if version = params["version"] do
  "ios/" <> version
end

File: test/t_web/channels/user_socket_test.exs [hunk 7]
```diff
@@ -3,18 +3,41 @@ defmodule TWeb.UserSocketTest do
...
-    test "with valid token", %{user: user} do
+    test "with valid token and no version", %{user: user} do
       token =
         user
         |> Accounts.generate_user_session_token("mobile")
         |> Accounts.UserToken.encoded_token()
 
       assert {:ok, _socket} = connect(UserSocket, %{"token" => token}, %{})
+
+      token = Accounts.UserToken |> where(user_id: ^user.id) |> Repo.one()
```
Comment (ruslandoga) [inline test/t_web/channels/user_socket_test.exs:21]:
```elixir
%Accounts.UserToken{version: version} = Accounts.UserToken |> where(user_id: ^user.id) |> Repo.one()
```

File: test/t_web/channels/user_socket_test.exs [hunk 8]
```diff
@@ -3,18 +3,41 @@ defmodule TWeb.UserSocketTest do
...
+
+    test "with valid token and version", %{user: user} do
+      token =
+        user
+        |> Accounts.generate_user_session_token("mobile")
+        |> Accounts.UserToken.encoded_token()
+
+      assert {:ok, _socket} = connect(UserSocket, %{"token" => token, "version" => "2.2.2"}, %{})
+
+      token = Accounts.UserToken |> where(user_id: ^user.id) |> Repo.one()
```
Comment (ruslandoga) [inline test/t_web/channels/user_socket_test.exs:36]: same as above

**Review by ruslandoga — APPROVED** [review]

**Pull request merged**